src/terms/xc.jl

```diff
@@ -259,53 +251,50 @@ function libxc_symmetric_index(i, j)
 end
 
 
+# stores the input to libxc in a format it likes
 struct LibxcDensity
     basis
     max_derivative::Int
-    ρ_real    # density on real-space grid
-    ∇ρ_real   # density gradient on real-space grid
-    σ_real    # contracted density gradient on real-space grid
+    ρ_real    # density ρ[iσ, ix, iy, iz]
+    ∇ρ_real   # for GGA, density gradient ∇ρ[iσ, ix, iy, iz, iα]
+    σ_real    # for GGA, contracted density gradient σ[iσ, ix, iy, iz]
 end
 
 """
 Compute density in real space and its derivatives starting from ρ
 """
-function LibxcDensity(basis, max_derivative::Integer, ρ::RealFourierArray, ρspin=nothing)
+function LibxcDensity(basis, max_derivative::Integer, ρ)
     model = basis.model
     @assert model.spin_polarization in (:collinear, :none, :spinless)
     @assert max_derivative in (0, 1)
-    model.spin_polarization == :collinear && (@assert !isnothing(ρspin))
-    ifft(x) = real_checked(G_to_r(basis, clear_without_conjugate!(x)))
 
     n_spin    = model.n_spin_components
     σ_real    = nothing
     ∇ρ_real   = nothing
-    if model.spin_polarization == :collinear
-        @assert n_spin == 2
-        ρ_real    = similar(ρ.real,    n_spin, basis.fft_size...)
-        ρ_real[1, :, :, :] .= @. (ρ.real + ρspin.real) / 2
-        ρ_real[2, :, :, :] .= @. (ρ.real - ρspin.real) / 2
-
-        if max_derivative > 0
-            ρ_fourier = similar(ρ.fourier, n_spin, basis.fft_size...)
-            ρ_fourier[1, :, :, :] .= @. (ρ.fourier + ρspin.fourier) / 2
-            ρ_fourier[2, :, :, :] .= @. (ρ.fourier - ρspin.fourier) / 2
+
+    # compute ρ_real and possibly ρ_fourier
+    ρ_real = similar(ρ, n_spin, basis.fft_size...)
+    for σ = 1:n_spin
+        ρ_real[σ, :, :, :] = ρ[:, :, :, σ]
```

Comment:
permutedims?

Reply:
would be a nice way to do it but less clear: this way the index order is clear for the rest of the code

Reply:
yes I agree that's a fair point. On the other hand if you put permutedims and a comment it's too.